src/main/java/hudson/remoting/Launcher.java

```diff
@@ -226,9 +263,46 @@ public static void main(String... args) throws Exception {
 
     @edu.umd.cs.findbugs.annotations.SuppressWarnings("DM_DEFAULT_ENCODING")    // log file, just like console output, should be in platform default encoding
     public void run() throws Exception {
-        if (slaveLog!=null) {
+
+        // Create and verify working directory if required
+        @CheckForNull Path internalDirPath = workDir != null ? workDir.toPath() : null;
+        if (internalDirPath == null) {
+            System.err.println("WARNING: Agent working directory is not specified. Some functionality introduced in Remoting 3 may be disabled");
+        } else {
+            if (workDir.exists()) {
+                if (!workDir.isDirectory()) {
+                    throw new IOException("The specified agent working directory path points to a non-directory file");
+                }
+                if (!workDir.canWrite() || !workDir.canRead() || !workDir.canExecute()) {
+                    throw new IOException("The specified agent working directory should be fully accessible to the remoting executable (RWX)");
+                }
+            }
+
+            // Now we create a subdirectory for remoting operations
+            internalDirPath = new File(workDir, internalDir).toPath();
+            Files.createDirectories(internalDirPath);
+            System.out.println("Using " + internalDirPath + " as a remoting working files directory");
+        }
+
+        if (slaveLog != null) { // Legacy behavior
+            System.out.println("Using " + slaveLog + " as an agent Error log destination. 'Out' log won't be generated");
+            System.out.flush(); // Just in case the channel
+            System.err.flush();
             System.setErr(new PrintStream(new TeeOutputStream(System.err,new FileOutputStream(slaveLog))));
+        } else if (internalDirPath != null) { // New behavior
+            System.out.println("Both error and output logs will be printed to " + internalDirPath);
+            System.out.flush();
+            System.err.flush();
+
+            // TODO: Log rotation by default?
+            System.setErr(new PrintStream(new TeeOutputStream(System.err,
+                    new FileOutputStream(new File(internalDirPath.toFile(), "remoting.err.log")))));
+            System.setOut(new PrintStream(new TeeOutputStream(System.out,
+                    new FileOutputStream(new File(internalDirPath.toFile(), "remoting.out.log")))));
```

Comment:
I'm wondering if this wouldn't be easier, for diagnostics purposes, to just output in a single file, and use JUL (I don't like it very much, but as it's already present in the JDK and Jenkins code already uses it in general...) or another logging API to distinguish between messages severities. E.g. here we don't have the dates of the logs, so it would make btw it hard or impossible to correlate .err logs to .out ones, right?